_source/_posts/2022-06-14-simple-crud-react-and-spring-boot.md

```diff
@@ -0,0 +1,1348 @@
+}
+
+@PostMapping("/group")
+ResponseEntity<Group> createGroup(@Valid @RequestBody Group group,
+                                  @AuthenticationPrincipal OAuth2User principal) throws URISyntaxException {
+    log.info("Request to create group: {}", group);
+    Map<String, Object> details = principal.getAttributes();
+    String userId = details.get("sub").toString();
+
+    // check to see if user already exists
+    Optional<User> user = userRepository.findById(userId);
+    group.setUser(user.orElse(new User(userId,
+                    details.get("name").toString(), details.get("email").toString())));
+
+    Group result = groupRepository.save(group);
+    return ResponseEntity.created(new URI("/api/group/" + result.getId()))
+            .body(result);
+}
+```
+
+## Modify React to handle CSRF and be identity-aware
+
+You'll need to make a few changes to your React components to make them identity-aware. The first thing you'll want to do is modify `src/index.js` to wrap everything in a `CookieProvider`. This component allows you to read the CSRF cookie and send it back as a header.
+
+```jsx
+import { CookiesProvider } from 'react-cookie';
+
+const root = ReactDOM.createRoot(document.getElementById('root'));
+root.render(
+  <React.StrictMode>
+    <CookiesProvider>
+      <App />
+    </CookiesProvider>
+  </React.StrictMode>
+);
+```
+
+Modify `app/src/Home.js` to call `/api/user` to see if the user is logged in. If they're not, show a `Login` button.
+
+```jsx
+import React, { useEffect, useState } from 'react';
+import './App.css';
+import AppNavbar from './AppNavbar';
+import { Link } from 'react-router-dom';
+import { Button, Container } from 'reactstrap';
+import { useCookies } from 'react-cookie';
+
+const Home = () => {
+
+  const [authenticated, setAuthenticated] = useState(false);
+  const [loading, setLoading] = useState(false);
+  const [user, setUser] = useState(undefined);
+  const [cookies] = useCookies(['XSRF-TOKEN']);
+
+  useEffect(() => {
+    setLoading(true);
+    fetch('api/user', { credentials: 'include' })
+      .then(response => response.text())
+      .then(body => {
+        if (body === '') {
+          setAuthenticated(false);
+        } else {
+          setUser(JSON.parse(body));
+          setAuthenticated(true);
+        }
+        setLoading(false);
+      });
+  }, [setAuthenticated, setLoading, setUser])
+
+  const login = () => {
+    let port = (window.location.port ? ':' + window.location.port : '');
+    if (port === ':3000') {
+      port = ':8080';
+    }
+    window.location.href = '//' + window.location.hostname + port + '/private';
+  }
+
+  const logout = () => {
+    fetch('/api/logout', {
+      method: 'POST', credentials: 'include',
+      headers: { 'X-XSRF-TOKEN': cookies['XSRF-TOKEN'] }
+    })
+      .then(res => res.json())
+      .then(response => {
+        window.location.href = response.logoutUrl + "?id_token_hint=" +
```

Comment:
```suggestion
        window.location.href = `${response.logoutUrl}?id_token_hint=${response.idToken}&post_logout_redirect_uri=${window.location.origin}`;
```

Reply:
Replaced with the following so there's not as much scrolling to the right.

```js
window.location.href = `${response.logoutUrl}?id_token_hint=${response.idToken}`
  + `&post_logout_redirect_uri=${window.location.origin}`;
```